File: src/main/java/com/kishlaly/ta/analyze/testing/sl/StopLossVolatileLocalMin.java
package com.kishlaly.ta.analyze.testing.sl;

import com.kishlaly.ta.model.QuoteJava;
import com.kishlaly.ta.model.SymbolDataJava;

import java.util.Comparator;

/**
 * SL is selected by {config} below the lowest quote.low of the {QUOTES_TO_FIND_MIN} bars before the current quote
 */
public class StopLossVolatileLocalMin extends StopLossStrategyJava {

    public static int QUOTES_TO_FIND_MIN = 20;

    public StopLossVolatileLocalMin(Object config) {
        super(config, true);
    }

    @Override
    public double calculate(SymbolDataJava data, int currentQuoteIndex) {
        QuoteJava quoteWithMinimalLow = data.quotes.subList(currentQuoteIndex - QUOTES_TO_FIND_MIN, currentQuoteIndex).stream().min(Comparator.comparingDouble(quote -> quote.getLow())).get();
        double distance = (double) config;
        return quoteWithMinimalLow.getLow() - distance;
    }

    @Override
    public String toString() {
        return "SL volatile local min";
    }
}

File: src/main/kotlin/com/kishlaly/ta/analyze/testing/sl/StopLossVolatileLocalMin.kt
package com.kishlaly.ta.analyze.testing.sl

import com.kishlaly.ta.model.SymbolData

class StopLossVolatileLocalMin(config: Any?) : StopLossStrategy(config, true) {

    companion object {
        var QUOTES_TO_FIND_MIN = 20
    }

    override fun calculate(data: SymbolData, signalIndex: Int): Double {
        val quoteWithMinimalLow = data.quotes.subList(signalIndex - QUOTES_TO_FIND_MIN, signalIndex).minBy { it.low }
        return quoteWithMinimalLow.low - config as Double
    }

    override fun toString() = "SL volatile local min"

}